Collect test names in lcov reports

jeremybettis opened the pull request:
If the flag `--coverage-by-instance` is passed to twister, pass the test name to lcov so that the coverage reports will include test details.

Here is an example without the test names populated:
![42AEQv8ZynYtoK7](https://github.com/user-attachments/assets/59219e84-c8dd-4535-b136-089026dab7af)

And with:
![w4t2fuyT64kWDRy](https://github.com/user-attachments/assets/973e6f8b-9b3b-4506-9844-a77f47d4fa93)

jeremybettis commented:
Ran twister on 6448 test cases

With `-coverage-by-testsuite`

```
real    102m34.790s
user    4588m59.765s
sys     415m0.437s
```

Without `-coverage-by-testsuite`

```
real    87m56.910s
user    4126m7.868s
sys     320m30.187s
```

So not a lot faster, but some.

golowanow commented:
@jeremybettis, nice to see you are also going in the same direction as in #66345 (RFC #67058)

From my understanding, this PR's `--coverage-by-testsuite` mode is similar to `--coverage-split` part of #66345 
but this PR doesn't make summary report for the default gcovr, e.g. 
```
twister -p unit_testing  --coverage  --coverage-by-testsuite -T tests/unit
...
ERROR   - Merge coverage not implemented for <twisterlib.coverage.Gcovr object at 0x7fbf68a75e80>
INFO - Saving reports...
INFO - Run completed
```

> .. This should help with parallelism, 

well, when I try it looks opposite for me (~10% slower), e.g.:
```
time ./scripts/twister -vv -p unit_testing  --coverage --coverage-tool lcov --coverage-by-testsuite -T tests/unit 
	real    1m32.694s   
	user    14m35.421s  
	sys     1m17.638s   

time ./scripts/twister -vv -p unit_testing  --coverage --coverage-tool lcov -T tests/unit
	real    1m19.423s   
	user    11m26.782s  
	sys     1m8.083s    
```

nashif commented:
> Also, the coverage reports will include test details.

this is nice, didn't know lcov can do that.

jeremybettis commented:
I'm resolving the merge conflicts, don't look at this PR.

jeremybettis committed:
twister: Fix --disable-coverage-aggregation help text

There is no flag named --coverage-split, switch help to
--coverage-per-instance

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis committed:
twister: Implement coverage file merging for lcov

Instead of capturing the coverage data twice with
--coverage-per-instance, just merge the lcov files in the aggregation
report.

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis committed:
twister: Collect test names in lcov

When using --coverage-per-instance, collect the test names in lcov, and
add --show-details to the html report to display those test names.

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis committed:
twister: Check lcov exit codes

Don't ignore lcov exit codes, but instead log and return if one of the
lcov commands fails.

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis commented:
Ok, I've rebased, and I went ahead and split up my change into 4 commits, and they are all very small.

jeremybettis commented on scripts/pylib/twister/twisterlib/coverage.py:
```diff
@@ -279,37 +279,75 @@ def _generate(self, outdir, coveragelog, build_dirs=None):
-            self.run_lcov(cmd, coveragelog)
+            ret = self.run_lcov(cmd, coveragelog)
+            if ret:
+                logger.error("LCOV remove ignores report stage failed with %s", ret)
```
No, the "stage" is that is is removing ignored files.

jeremybettis commented on scripts/pylib/twister/twisterlib/coverage.py:
```diff
@@ -325,7 +329,10 @@ def _generate(self, outdir, coveragelog, build_dirs=None):
-               "-output-directory", os.path.join(outdir, "coverage")] + files
+               "-output-directory", os.path.join(outdir, "coverage")]
+        if self.coverage_per_instance:
+            cmd.append("--show-details")
```
Do the tests have descriptions?
